src/VisualStudio/Core/Def/Implementation/ChangeSignature/ChangeSignatureDialogViewModel.ParameterViewModels.cs

```diff
@@ -0,0 +1,246 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Linq;
+using System.Windows;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ChangeSignature;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ChangeSignature
+{
+    internal partial class ChangeSignatureDialogViewModel
+    {
+        public abstract class ParameterViewModel
+        {
+            protected ChangeSignatureDialogViewModel ChangeSignatureDialogViewModel { get; }
+
+            public abstract Parameter Parameter { get; }
+
+            public abstract string Type { get; }
+            public abstract string ParameterName { get; }
+            public abstract bool IsRemoved { get; set; }
+            public abstract bool IsDisabled { get; }
+            public abstract string CallSite { get; }
+            public abstract bool IsRequired { get; }
+            public abstract string DefaultValue { get; }
+
+            public abstract string InitialIndex { get; }
+            public abstract string Modifier { get; }
+            public abstract string Default { get; }
+
+            public ParameterViewModel(ChangeSignatureDialogViewModel changeSignatureDialogViewModel)
+            {
+                ChangeSignatureDialogViewModel = changeSignatureDialogViewModel;
+            }
+
+            public Visibility HasParameterNameConflict { get; set; }
+
+            public virtual Visibility TypeWarningVisibility => Visibility.Collapsed;
+
+            public string ShortAutomationText => $"{Type} {ParameterName}";
+
+            public virtual string FullAutomationText
+            {
+                get
+                {
+                    var text = $"{Modifier} {Type} {Parameter}";
+                    if (!string.IsNullOrWhiteSpace(Default))
+                    {
+                        text += $" = {Default}";
+                    }
+
+                    return text;
+                }
+            }
+
+            public bool NeedsBottomBorder
```

Comment:
hmm... this design seems a little odd. Could the be abstract as well and handled in the implementing classes?

Reply:
Why would the implementing classes handle this? The logic is the same for Existing/Added?

Reply:
The `this ==` is basically checking implementation details of subclasses right?

Reply:
Talked offline. Going to leave this as-is.